common/Reducers.jsm

```diff
@@ -457,6 +457,182 @@ function Pocket(prevState = INITIAL_STATE.Pocket, action) {
+              return Object.assign({}, item, {
+                bookmarkGuid,
+                bookmarkTitle,
+                bookmarkDateCreated: dateAdded,
+              });
+            }
+            return item;
+          }),
+        } : {},
+      };
+    default:
+      return state;
+  }
+}
+
+// Update recommendations data inside discovery stream feeds by handling
+// action data received on click of link menu options
+function DiscoveryStreamFeeds(state = {}, action) {
+  switch (action.type) {
+    case at.PLACES_LINK_BLOCKED:
+      return {
+        ...state.feeds,
+        data: Object.keys(state.feeds.data).reduce((accumulator, feed_url) => {
+          accumulator[feed_url] = {
+            data: {
+              ...state.feeds.data[feed_url].data,
+              recommendations: state.feeds.data[feed_url].data.recommendations.filter(r => r.url !== action.data.url),
+            },
+          };
+          return accumulator;
+        }, {}),
+      };
+    // Update item by adding pocket_id from action.data.pocket_id
+    case at.PLACES_SAVED_TO_POCKET:
+      return {
+        ...state.feeds,
+        data: Object.keys(state.feeds.data).reduce((accumulator, feed_url) => {
+          accumulator[feed_url] = {
+            data: {
+              ...state.feeds.data[feed_url].data,
+              recommendations: state.feeds.data[feed_url].data.recommendations.map(item => {
+                if (item.url === action.data.url) {
+                  return Object.assign({}, item, {
+                    open_url: action.data.open_url,
+                    pocket_id: action.data.pocket_id,
+                  });
+                }
+                return item;
+              }),
+            },
+          };
+          return accumulator;
+        }, {}),
+      };
+    case at.DELETE_FROM_POCKET:
+    case at.ARCHIVE_FROM_POCKET:
+      return {
+        ...state.feeds,
+        data: Object.keys(state.feeds.data).reduce((accumulator, feed_url) => {
+          accumulator[feed_url] = {
+            data: {
+              ...state.feeds.data[feed_url].data,
+              recommendations: state.feeds.data[feed_url].data.recommendations.filter(r => r.pocket_id !== action.data.pocket_id),
+            },
+          };
+          return accumulator;
+        }, {}),
+      };
+    // Update item inside feeds by adding bookmarkGuid from action.data.bookmarkGuid
+    case at.PLACES_BOOKMARK_ADDED:
+      return {
+        ...state.feeds,
+        data: Object.keys(state.feeds.data).reduce((accumulator, feed_url) => {
+          accumulator[feed_url] = {
+            data: {
+              ...state.feeds.data[feed_url].data,
+              recommendations: state.feeds.data[feed_url].data.recommendations.map(item => {
+                if (item.url === action.data.url) {
+                  const {bookmarkGuid, bookmarkTitle, dateAdded} = action.data;
+                  return Object.assign({}, item, {
+                    bookmarkGuid,
```

Comment:
Just a quick glance.. this is the only place bookmarkGuid is set, i.e., when the bookmark is added? This won't persist after restarting firefox as that would probably want DiscoveryStreamFeed changes to query every single url, but this loss-on-restart behavior might be acceptable for the initial implementation and have a bug for later/backlog ?

Reply:
we have https://bugzilla.mozilla.org/show_bug.cgi?id=1531596 for items losing bookmarked state on restart. Default Activity Stream top stories are keeping bookmarked state on page refresh, will look and weigh in effort needed to port that support